streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/core/master/resourcemanager/ResourceManagerImpl.java

```diff
@@ -0,0 +1,205 @@
+
+    this.slotAssignStrategy = SlotAssignStrategyFactory.getStrategy(slotAssignStrategyType);
+    this.slotAssignStrategy.setResources(resources);
+    LOG.info("Slot assign strategy: {}.", slotAssignStrategy.getName());
+
+    checkAndUpdateResources();
+
+    this.scheduledExecutorService = Executors.newScheduledThreadPool(1);
+    long intervalSecond = resourceConfig.resourceCheckIntervalSecond();
+    this.scheduledExecutorService.scheduleAtFixedRate(
+        Ray.wrapRunnable(this::checkAndUpdateResources), 0, intervalSecond, TimeUnit.SECONDS);
+
+    LOG.info("ResourceManagerImpl init success.");
+  }
+
+  @Override
+  public Map<String, Double> allocateResource(final ExecutionVertex executionVertex) {
+    //executionVertex -> slot -> container
+    Container container = resources.getRegisterContainerByContainerId(
+        executionVertex.getSlot().getContainerID());
+    LOG.info("Start to allocate resource for actor with container: {}.", container);
+
+    // allocate resource to actor
+    Map<String, Double> resources = new HashMap<>();
+    resources.put(container.getName(), 1.0);
+
+    Map<String, Double> containResource = container.getAvailableResource();
+    for (Map.Entry<String, Double> entry : containResource.entrySet()) {
+      Map<String, Double> requireResource = executionVertex.getResources();
+      if (requireResource.containsKey(entry.getKey())) {
+        double availableResource = entry.getValue() - requireResource.get(entry.getKey());
+        entry.setValue(availableResource);
+      }
+    }
+
+    LOG.info("Allocate resource to actor [vertexId={}] succeeded with container {}.",
+        executionVertex.getVertexId(), container);
+    return resources;
+  }
+
+  @Override
+  public void deallocateResource(final ExecutionVertex executionVertex) {
+    LOG.info("Start deallocate resource for actor {}.", executionVertex.getWorkerActorId());
+    Container container = resources.getRegisterContainerByContainerId(
+        executionVertex.getSlot().getContainerID());
+
+    Map<String, Double> containResource = container.getAvailableResource();
+    for (Map.Entry<String, Double> entry : containResource.entrySet()) {
+      Map<String, Double> requireResource = executionVertex.getResources();
+      if (requireResource.containsKey(entry.getKey())) {
+        double availableResource = entry.getValue() + requireResource.get(entry.getKey());
+        entry.setValue(availableResource);
+      }
+    }
+
+    LOG.info("Deallocate resource for actor {} success.", executionVertex.getWorkerActorId());
+  }
+
+  @Override
+  public List<Container> getRegisteredContainers() {
+    return new ArrayList<>(resources.getRegisterContainers());
+  }
+
+  @Override
+  public SlotAssignStrategy getSlotAssignStrategy() {
+    return slotAssignStrategy;
+  }
+
+  @Override
+  public Resources getResources() {
+    return this.resources;
+  }
+
+  private void checkAndUpdateResources() {
```

Comment:
Although the code tells the purpose of what the method want to do, it might be great to add description for this method, because the calculation of the complement set will take some time to understand.

Reply:
added